### ISSUE TITLE
Constructor should not include "total" in list of items

### ISSUE THREAD

**Issue by halindrome** [WICG/paymentrequest#68]
Right now the "total" cost of the transaction is reflected as an item in the items sequence of the details object.  This is an anti-pattern.  The total should be a separate item in that object (and, in fact, there should be no requirement for items - that is optional detail about the transaction).

**Comment by zkoch** [comment]
Can you extrapolate more on why it's an anti pattern?

**Comment by webpayments** [comment]
Sure.  Positional parameters are an anti-pattern.  It is why you use
objects with named elements in modern programming instead of sending an
array where item 0 is the first name and item 1 is the last name.  It
doesn't scale and it assumes arcane knowledge.  A list of items belows in
an item list.  Tax and shipping costs could arguably be called items...  I
don't agree but I am not really willing to fight about it.  I would put
them in a different list (e.g., fees or something).  The total is neither
an item nor a fee.  It is the total.  It is the only thing that really
matters.  And it shouldn't be discoverable by needing to know that it is
the last item in a list.

On Wed, Feb 24, 2016 at 3:07 PM, Zach Koch <notifications@github.com> wrote:

> Can you extrapolate more on why it's an anti pattern?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/68#issuecomment-188503403>.
>



-- 
Shane McCarron
Projects Manager, Spec-Ops

**Comment by dezell** [comment]
"A man with two watches never knows what time it is."  If it is possible to sum items and come up with a different number than what is in total, that begs a question - why?  This effect can be mitigated by (e.g.) making the total the authoritative number, and make it clear that the rest is merely informative, like comments.

**Comment by dlongley** [comment]
I don't know how authoritative total can be here. It all seems informative to me (to simply improve UX).

Only once the site has collected everything they need can it generate the payment request message that will be forwarded to the payment app. It's in the payment app that the user will actually consent to paying, not in the browser. So it's that information (which may only have a total, no line items) that is authoritative.

**Comment by halindrome** [comment]
Umm...  This *is* the message to the payment app.  At least in that I expect all of this data is supposed to be passed there.  Otherwise the payment app cannot fulfill its relationship with the PSP by, for example, passing on the airline itinerary or whatever (which would be in some additional attributes attached to the object).  I assume all of this data modulo the things outside of the intersection of the merchant and payment app supported payment methods would be passed to the payment app.

**Comment by halindrome** [comment]
@zkoch - the comment of mine you could not find is above labelled as from "webpayments".  I guess I did it in email instead of in here.  My bad!

**Comment by nickjshearer** [comment]
> The total is neither an item nor a fee.  It is the total.  It is the only thing that really
matters.

In the spirit of stimulating further discussion, I shall disagree with that point. Sometimes from the user's perspective the total is the *last* thing that matters. This is because some types of payment instrument - such as an EMV card - require an absolute amount to be charged (the 'total'). However, this absolute amount may have little to no bearing on the expected cost of the product (an Uber or Lyft pre-auth, for example).

From the payment provider's perspective, the total may well be the only thing that really matters. That is not true from the user's perspective.

**Comment by webpayments** [comment]
Apologies - I of course meant "the only thing that really matters to the
payment flow".  The user needs to know all sorts of stuff.  I am not
proposing that we get rid of the items list.  I am proposing that the total
be a separate item - in particular because it is by definition NOT an item.

On Fri, Feb 26, 2016 at 9:34 AM, Nick S <notifications@github.com> wrote:

> The total is neither an item nor a fee. It is the total. It is the only
> thing that really
> matters.
>
> In the spirit of stimulating further discussion, I shall disagree with
> that point. Sometimes from the user's perspective the total is the *last*
> thing that matters. This is because some types of payment instrument - such
> as an EMV card - require an absolute amount to be charged (the 'total').
> However, this absolute amount may have little to no bearing on the expected
> cost of the product (an Uber or Lyft pre-auth, for example).
>
> From the payment provider's perspective, the total may well be the only
> thing that really matters. That is not true from the user's perspective.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WICG/paymentrequest/issues/68#issuecomment-189384371>.
>



-- 
Shane McCarron
Projects Manager, Spec-Ops

**Comment by rsolomakhin** [comment]
@halindrome , could you give a code example of how this would look? Then we can talk about something more concrete, imho.

**Comment by halindrome** [comment]
Of course - I was just waiting for the repo to move.  I thought I saw mail that was happening and we should hold off on PRs.  Maybe I just dreamed it.

**Comment by zkoch** [comment]
Yeah, PRs are on hold just for a bit longer while the migration happens.

**Comment by adrianba** [comment]
Now tracked at w3c/browser-payment-api#18.
